modulemd/v2/modulemd-module.c

```diff
@@ -181,23 +181,68 @@ modulemd_module_init (ModulemdModule *self)
 }
 
 
-void
-modulemd_module_set_defaults (ModulemdModule *self, ModulemdDefaults *defaults)
+ModulemdDefaultsVersionEnum
+modulemd_module_set_defaults (ModulemdModule *self,
+                              ModulemdDefaults *defaults,
+                              ModulemdDefaultsVersionEnum index_mdversion,
+                              GError **error)
 {
-  g_return_if_fail (MODULEMD_IS_MODULE (self));
+  g_autoptr (ModulemdDefaults) upgraded_defaults = NULL;
+  g_autoptr (GError) nested_error = NULL;
+  g_return_val_if_fail (MODULEMD_IS_MODULE (self), MD_DEFAULTS_VERSION_ERROR);
 
   g_clear_object (&self->defaults);
-
   if (defaults == NULL)
-    return;
+    {
+      /* If we are empty here, return the passed-in index_mdversion so the
+       * function reports success and does not influence further upgrades.
+       */
+      return index_mdversion;
+    }
 
   /* We should never get a defaults object added whose module name doesn't
    * match.
    */
-  g_return_if_fail (g_str_equal (modulemd_defaults_get_module_name (defaults),
-                                 modulemd_module_get_module_name (self)));
+  if (g_strcmp0 (modulemd_defaults_get_module_name (defaults),
+                 modulemd_module_get_module_name (self)) != 0)
+    {
+      g_set_error (error,
+                   MODULEMD_ERROR,
+                   MODULEMD_ERROR_VALIDATE,
+                   "Attempted to add defaults for module '%s' to module '%s'",
+                   modulemd_defaults_get_module_name (defaults),
+                   modulemd_module_get_module_name (self));
+      return MD_DEFAULTS_VERSION_ERROR;
+    }
+
+  /* For Modulemd.ModuleIndex, we always want all entries to have the same
+   * version, so that merges can be performed. If this Defaults object has a
+   * lower mdversion than the Index, upgrade it to that version.
+   */
+  if (modulemd_defaults_get_mdversion (defaults) < index_mdversion)
```

Comment:
Note that the `modulemd_defaults_upgrade` implementation already does the copy if the mdversion is equal to the requested version, so you could simplify here by just always calling the upgrade function.

Reply:
No, I can't. I'll update the comment to explain why, but it's basically this: if the version we have here is already *higher* than the index_mdversion, the upgrade routine would fail because it would be asked to downgrade.